### PR TITLE
cleanup of linear/svc

### LINKED ISSUE #125: Cleanup Linear Converter Code
We have the original draft of the linear code uploaded in [this branch](https://github.com/microsoft/hummingbird/tree/cleanup/linear-converter). All credit here goes to @scnakandala, the original author and brains behind this.

It contains an un-edited implementation of the following scikit-learn converters:
* [linear_classifier.py](https://github.com/microsoft/hummingbird/blob/cleanup/linear-converter/hummingbird/ml/operator_converters/linear_classifier.py)
  * LinearRegression
  * LogisticRegression
  * LinearSVC
  * SGDClassifier
  * LogisticRegressionCV

* [svc.py](https://github.com/microsoft/hummingbird/blob/cleanup/linear-converter/hummingbird/ml/operator_converters/svc.py)
  * SVC
  * NuSVC

There is a single test file [here](https://github.com/microsoft/hummingbird/blob/cleanup/linear-converter/tests/test_sklearn_linear_converters.py) that needs to be cleaned up and separated out.

### PR DESCRIPTION
**Pull request by ksaur**

### CODE REVIEW

File: hummingbird/ml/operator_converters/linear_classifier.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# -------------------------------------------------------------------------
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/linear_classifier.py]: Should we call the file `linear` rather then `linear_classifier`? We do regression, not just classification.

File: hummingbird/ml/operator_converters/linear_classifier.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        self.perform_class_select = False
+        if min(classes) != 0 or max(classes) != len(classes) - 1:
+            self.perform_class_select = True
+
+        self.binary_classification = False
+        if len(classes) == 2:
+            self.binary_classification = True
+
+    def forward(self, x):
+        if self.multi_class == "multinomial":
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/linear_classifier.py]: Can we pull `torch.addmm(self.intercepts, x, self.coefficients)` out put it in the first line? Then `if self.regression` will just return output.

File: hummingbird/ml/operator_converters/linear_classifier.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        else:
+            return torch.argmax(output, dim=1), output
+
+
+def convert_sklearn_linear_model(operator, device, extra_config):
+    """
+    Converter for `sklearn.svm.LinearSVC`, `sklearn.linear_model.LinearRegression`,
+        `sklearn.linear_model.LogisticRegression`, `sklearn.linear_model.SGDClassifier`, and
+        `sklearn.linear_model.LogisticRegressionCV`
+
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/linear_classifier.py:59]: Remove one line.

File: hummingbird/ml/operator_converters/linear_classifier.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        if self.binary_classification:
+            output = torch.cat([1 - output, output], dim=1)
+
+        if self.perform_class_select:
+            return torch.index_select(self.classes, 0, torch.argmax(output, dim=1)), output
+        else:
+            return torch.argmax(output, dim=1), output
+
+
+def convert_sklearn_linear_model(operator, device, extra_config):
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/linear_classifier.py]: should we have a different method for linear regression? It looks that the code for linear regression is quite different than the other operators. In this way we can avoid to have to understand which branch will be taken and just explicitly spell out the conversion.

File: tests/test_sklearn_linear_converters.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        X = np.random.rand(100, 200)
+        X = np.array(X, dtype=np.float32)
+        y = np.random.randint(num_classes, size=100) + labels_shift
+
+        model.fit(X, y)
+
+        pytorch_model = hummingbird.ml.convert(model, "pytorch")
+
+        self.assertTrue(pytorch_model is not None)
+        np.testing.assert_allclose(model.predict_proba(X), pytorch_model.predict_proba(X), rtol=1e-5, atol=1e-6)
```
Comment (interesaaat) [inline tests/test_sklearn_linear_converters.py]: Should both be 1e-06?

Comment (interesaaat) [reply tests/test_sklearn_linear_converters.py]: Same below.

File: tests/test_sklearn_svc.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
+"""
```
Comment (interesaaat) [inline tests/test_sklearn_svc.py]: Can we call the file `test_sklearn_sv`? We will add regression sooner or later.

File: hummingbird/ml/operator_converters/svc.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# -------------------------------------------------------------------------
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/svc.py]: Can we call the file `sv`? We will add regression sooner or later.

File: hummingbird/ml/supported.py [hunk 8]
```diff
@@ -64,6 +77,14 @@ def _build_sklearn_operator_list():
             HistGradientBoostingRegressor,
             RandomForestClassifier,
             RandomForestRegressor,
+            # Linear-methods
+            LinearRegression,
+            LinearSVC,
+            LogisticRegression,
+            LogisticRegressionCV,
+            NuSVC,
```
Comment (interesaaat) [inline hummingbird/ml/supported.py]: can we put `NuSVC` and `SVC` together and under a `#SV` comment? (They are not linear methods)

File: tests/test_sklearn_svc.py (hunk 6 above)
Comment (ksaur) [reply tests/test_sklearn_svc.py]: you mean `test_sklearn_sv`

File: hummingbird/ml/operator_converters/linear_classifier.py (hunk 1 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/linear_classifier.py]: done

File: tests/test_sklearn_linear_converters.py (hunk 5 above)
Comment (ksaur) [reply tests/test_sklearn_linear_converters.py]: done

File: hummingbird/ml/supported.py (hunk 8 above)
Comment (ksaur) [reply hummingbird/ml/supported.py]: done

File: hummingbird/ml/operator_converters/linear_classifier.py (hunk 4 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/linear_classifier.py]: please take another look? ok now?

File: hummingbird/ml/operator_converters/linear_classifier.py (hunk 2 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/linear_classifier.py]: @interesaaat  - ok, please take a look and make sure this didn't break anything.

Mmmmm paper deadline code from December. delicious.

File: hummingbird/ml/operator_converters/skl_linear.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        self.binary_classification = False
+        if len(classes) == 2:
+            self.binary_classification = True
+
+    def forward(self, x):
+        output = torch.addmm(self.intercepts, x, self.coefficients)
+        if self.multi_class == "multinomial":
+            output = torch.softmax(output, dim=1)
+        elif self.regression:
+            if not self.binary_classification:
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_linear.py]: This can be removed. Just return `output`.

File: hummingbird/ml/operator_converters/skl_linear.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        if self.perform_class_select:
+            return torch.index_select(self.classes, 0, torch.argmax(output, dim=1)), output
+        else:
+            return torch.argmax(output, dim=1), output
+
+
+def convert_sklearn_linear_model(operator, device, extra_config):
+    """
+    Converter for `sklearn.svm.LinearSVC`, `sklearn.linear_model.LogisticRegression`,
+        `sklearn.linear_model.SGDClassifier`, and `sklearn.linear_model.LogisticRegressionCV`
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_linear.py]: No need for tab here.

File: hummingbird/ml/operator_converters/skl_linear.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        device: String defining the type of device the converted operator should be run on
+        extra_config: Extra configuration used to select the best conversion strategy
+
+    Returns:
+        A PyTorch model
+    """
+    classes = [0] if not hasattr(operator.raw_operator, "classes_") else operator.raw_operator.classes_
+
+    if not all([type(x) in [int, np.int32, np.int64] for x in classes]):
+        raise RuntimeError("hummingbird supports only integer labels for class labels.")
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_linear.py]: `"Hummingbird currently supports only integer labels for class labels. Please file an issue at https://github.com/microsoft/hummingbird."

File: hummingbird/ml/operator_converters/skl_linear.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+    coefficients = operator.raw_operator.coef_.transpose().astype("float32")
+    intercepts = operator.raw_operator.intercept_.reshape(1, -1).astype("float32")
+
+    if hasattr(operator.raw_operator, "multi_class"):
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_linear.py:78]: Can we do something like:
`multi_class = None`
Then only one branch? I think it is cleaner.

File: hummingbird/ml/operator_converters/skl_linear.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        extra_config: Extra configuration used to select the best conversion strategy
+
+    Returns:
+        A PyTorch model
+    """
+
+    coefficients = operator.raw_operator.coef_.transpose().reshape(-1, 1).astype("float32")
+    intercepts = operator.raw_operator.intercept_.reshape(1, -1).astype("float32")
+
+    return SklearnLinearModel(coefficients, intercepts, [0], None, device, is_linear_regression=True)
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_linear.py]: Can we set these as default params in the class and move device after intercepts? In this way for linear regression we only need to write `return SklearnLinearModel(coefficients, intercepts, device)`.

File: hummingbird/ml/operator_converters/skl_sv.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+class SVC(torch.nn.Module):
+    def __init__(self, kernel, degree, sv, nv, a, b, gamma, coef0, classes, device):
+        super(SVC, self).__init__()
+        self.kernel = kernel
+        self.degree = degree
+        self.gamma = gamma
+        self.regression = False
+        sv = sv.toarray() if type(sv) == scipy.sparse.csr.csr_matrix else sv
+
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_sv.py]: Any specific reason why we have space here? Otherwise we can just remove it.

File: hummingbird/ml/supported.py [hunk 15]
```diff
@@ -64,6 +77,15 @@ def _build_sklearn_operator_list():
             HistGradientBoostingRegressor,
             RandomForestClassifier,
             RandomForestRegressor,
+            # Linear-methods
+            LinearRegression,
+            LinearSVC,
+            LogisticRegression,
+            LogisticRegressionCV,
+            SGDClassifier,
+            # SV
```
Comment (interesaaat) [inline hummingbird/ml/supported.py]: Sorry `SVM` here.

File: tests/test_sklearn_linear_converters.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        self._test_logistic_regression(3, multi_class="ovr")
+
+    # LogisticRegression with multi+multinomial
+    def test_logistic_regression_multi_multin(self):
+        warnings.filterwarnings("ignore")
+        # this will not converge due to small test size
+        self._test_logistic_regression(3, multi_class="multinomial", solver="sag")
+
+    # LinearRegression test function to be parameterized
+    def _test_linear_regression(self, y_input):
```
Comment (interesaaat) [inline tests/test_sklearn_linear_converters.py:58]: We should do the same also for the other regression tests.

File: hummingbird/ml/operator_converters/svc.py (hunk 7 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/svc.py]: done

File: hummingbird/ml/operator_converters/skl_sv.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            self.perform_class_select = True
+        self.n_classes = len(classes)
+
+    def forward(self, x):
+        x = x.double()
+
+        if self.kernel == "linear":
+            k = torch.mm(x, self.sv_t)
+        elif self.kernel == "rbf":
+            # using quadratic expansion--susseptible to rounding-off errors
```
Comment (interesaaat) [inline hummingbird/ml/operator_converters/skl_sv.py:51]: Can you please add a reference to [this](http://www.robots.ox.ac.uk/~albanie/notes/Euclidean_distance_trick.pdf) paper in the comment?

File: tests/test_sklearn_linear_converters.py (hunk 16 above)
Comment (ksaur) [reply tests/test_sklearn_linear_converters.py:58]: Which exactly? these (at least not all of these) tests don't make sense for LinearReg...  Coverage is complete; is there a specific sub-case you have in mind that is missing?  Or did you mean for SDG?

File: hummingbird/ml/operator_converters/skl_linear.py (hunk 9 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_linear.py]: done

File: hummingbird/ml/operator_converters/skl_linear.py (hunk 10 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_linear.py]: done

File: hummingbird/ml/operator_converters/skl_linear.py (hunk 11 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_linear.py]: done

File: hummingbird/ml/operator_converters/skl_linear.py (hunk 12 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_linear.py:78]: good  point

File: hummingbird/ml/operator_converters/skl_sv.py (hunk 14 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_sv.py]: the sv does seem a bit lost.

File: hummingbird/ml/supported.py (hunk 15 above)
Comment (ksaur) [reply hummingbird/ml/supported.py]: LO.
LOL.

File: hummingbird/ml/operator_converters/skl_sv.py (hunk 17 above)
Comment (ksaur) [reply hummingbird/ml/operator_converters/skl_sv.py:51]: yupper.

**Comment by ksaur** [comment]
I think it's ready, PTAL @interesaaat 

(to be merged  after #145 )
